viewer/packages/tools/src/SmartOverlay/SmartOverlayTool.ts

```diff
@@ -0,0 +1,400 @@
+        idsToRemove.push(overlayIndex);
+      }
+    });
+
+    this._singeOverlayCollection.removeOverlays(idsToRemove);
+  }
+
+  /**
+   * Removes overlays that were added with addOverlays method.
+   * @param id Id of the overlay group to remove.
+   */
+  removeOverlays(id: OverlayGroupId): void {
+    const { _viewer: viewer } = this;
+    const points = this._overlayPoints.find(points => points.uuid === id);
+    this._overlayPoints.filter(points => points.uuid !== id);
+
+    if (points) {
+      viewer.removeObject3D(points);
+      points.dispose();
+    }
+  }
+
+  /**
+   * Sets whether new overlays can be added.
+   */
+  set enabled(enabled: boolean) {
+    this._isEnabled = enabled;
+  }
+
+  get enabled(): boolean {
+    return this._isEnabled;
+  }
+
+  /**
+   * Sets whether overlays are visible.
+   */
+  set visible(visible: boolean) {
+    this._isVisible = visible;
+    this._overlayPoints.forEach(points => {
+      points.visible = visible;
+    });
+
+    this._viewer.requestRedraw();
+  }
+
+  get visible(): boolean {
+    return this._isVisible;
+  }
+
+  set textOverlayVisible(visible: boolean) {
+    this._textOverlayVisible = visible;
+    this._textOverlay.style.opacity = visible ? '1' : '0';
+  }
+
+  get textOverlayVisible(): boolean {
+    return this._textOverlayVisible;
+  }
+
+  /**
+   * Removes all overlays.
+   */
+  clear(): void {
+    this._singeOverlayCollection.removeAllOverlays();
+
+    this._overlayPoints.forEach(points => {
+      this._viewer.removeObject3D(points);
+      points.dispose();
+    });
+    this._overlayPoints.splice(0, this._overlayPoints.length);
+
+    this._labelIndexToLabelInfoMap.clear();
+  }
+
+  on(event: 'hover', eventHandler: OverlayEventHandler): void;
+  on(event: 'click', eventHandler: OverlayEventHandler): void;
+
+  on(event: OverlayToolEvent, eventHandler: OverlayEventHandler): void {
+    switch (event) {
+      case 'hover':
+        this._events.hover.subscribe(eventHandler);
+        break;
+      case 'click':
+        this._events.click.subscribe(eventHandler);
+        break;
+      default:
+        assertNever(event);
+    }
+  }
+
+  off(event: 'hover', eventHandler: OverlayEventHandler): void;
+  off(event: 'click', eventHandler: OverlayEventHandler): void;
+
+  off(event: OverlayToolEvent, eventHandler: OverlayEventHandler): void {
+    switch (event) {
+      case 'hover':
+        this._events.hover.unsubscribe(eventHandler);
+        break;
+      case 'click':
+        this._events.click.unsubscribe(eventHandler);
+        break;
+      default:
+        assertNever(event);
+    }
+  }
+
+  private positionTextOverlay(event: MouseEvent): void {
+    const { _textOverlay, _textOverlayVisible } = this;
+    _textOverlay.style.left = `${event.offsetX}px`;
+    _textOverlay.style.top = `${event.offsetY}px`;
+    _textOverlay.style.opacity = _textOverlayVisible ? '1' : '0';
+  }
+
+  private intersectPointsMarkers(mouseCoords: {
```

Comment:
How well does this scale? It will be executed ALL THE TIME, so if we keep this it needs to be SUPER-performant.

Reply:
It works for million of points, I was also surprised about the performance. What takes a lot of time is not the intersection actually but updating the scale of the Icon, so minimising number of icon updates saves a lot of performance. Optimisation can be done in the future but I think 1 million is good enough for now.

Reply:
Works ok-ish even for 2 million points, so I think we are good for first iteration. Quite obvious optimisation is actually using octree for raycasting, this should increase the speed by a lot but require some complexity to be added.